espresso/environment/optitmism_espresso_test_helpers.go

```diff
@@ -0,0 +1,507 @@
+}
+
+// FailedToLoadEspressoAccount represents a class of errors that occur when we
+// are unable to load the espresso account
+type FailedToLoadEspressoAccount struct {
+	Cause error
+}
+
+// Error implements error
+func (f FailedToLoadEspressoAccount) Error() string {
+	return fmt.Sprintf("failed to load the espresso account: %v", f.Cause)
+}
+
+// FailedToLaunchDockerContainer represents a class of errors that occur when
+// we are unable to launch a docker container
+type FailedToLaunchDockerContainer struct {
+	Cause error
+}
+
+// Error implements error
+func (f FailedToLaunchDockerContainer) Error() string {
+	return fmt.Sprintf("failed to launch docker container: %v", f.Cause)
+}
+
+// EspressoNodeFailedToBecomeReady represents a class of errors that indicate
+// that the espresso-dev-node failed to become ready.
+type EspressoNodeFailedToBecomeReady struct {
+	Cause error
+}
+
+// Error implements error
+func (e EspressoNodeFailedToBecomeReady) Error() string {
+	return fmt.Sprintf("espresso node failed to become ready: %v", e.Cause)
+}
+
+type EspressoDevNodeContainerInfo struct {
+	ContainerInfo DockerContainerInfo
+}
+
+var _ EspressoDevNode = (*EspressoDevNodeContainerInfo)(nil)
+
+func (e EspressoDevNodeContainerInfo) getPort(originalPort string) string {
+	hosts := e.ContainerInfo.PortMap[originalPort]
+
+	if len(hosts) == 0 {
+		return ""
+	}
+
+	_, port, err := net.SplitHostPort(hosts[0])
+	if err != nil {
+		return ""
+	}
+
+	return port
+}
+
+// SequencerPort implements EspressoDevNode, by returning the relevant
+// port for the sequencer API in the Espresso dev node
+func (e EspressoDevNodeContainerInfo) SequencerPort() string {
+	return e.getPort(ESPRESSO_SEQUENCER_API_PORT)
+}
+
+// BuilderPort implements EspressoDevNode, by returning the relevant
+// port for the builder API in the Espresso dev node
+func (e EspressoDevNodeContainerInfo) BuilderPort() string {
+	return e.getPort(ESPRESSO_BUILDER_PORT)
+}
+
+// Stop implements EspressoDevNode, and is a convenience method to stop the
+// running container.
+//
+// This is mostly unnecessary as the context that the container was launched
+// in will govern the lifecycle of the container automatically, assuming that
+// the context is following the recommended context usage patterns.
+func (e EspressoDevNodeContainerInfo) Stop() error {
+	cli := new(DockerCli)
+	return cli.StopContainer(context.Background(), e.ContainerInfo.ContainerID)
+}
+
+// ErrUnableToDetermineEspressoDevNodeSequencerHost is a sentinel error that
+// indicates that we were unable to determine what the Sequencer API host
+// is meant to be.
+var ErrUnableToDetermineEspressoDevNodeSequencerHost = errors.New("unable to determine the host for the espresso-dev-node sequencer api")
+
+func (l *EspressoDevNodeLauncherDocker) StartDevNet(ctx context.Context, t *testing.T, options ...DevNetLauncherOption) (*e2esys.System, EspressoDevNode, error) {
```

Comment:
Does this function allow to launch an OP node and a Caff node simultaneously?

Reply:
At the moment, this just launches the `op e2e` local devnet with the `espresso-dev-node` running.  This doesn't have anything to do with the `Caff-Node`.

Since the `Caff-Node` would be dependent on the `espresso-dev-node`, and the `L1`, I would imagine it would require this to be launched before being launched itself, and as a result, it would be separate from this process, but dependent on it.